qa/rpc-tests/addressindex.py

```diff
@@ -0,0 +1,389 @@
+
+            bal = self.nodes[i].getaddressbalance(b)
+            assert_equal(bal['received'], 0)
+            assert_equal(bal['balance'], 0)
+
+        # now re-mine the a to b send
+        self.nodes[0].generate(1)
+        self.sync_all()
+        for i in range(3):
+            assert_equal(self.nodes[i].getblockcount(), 111)
+
+        mempool = self.nodes[0].getaddressmempool({'addresses': [b, a]})
+        assert_equal(len(mempool), 0)
+
+        bal = self.nodes[2].getaddressbalance(a)
+        assert_equal(bal['received'], expected*COIN)
+        # the value 4 UTXO is no longer in our balance
+        assert_equal(bal['balance'], (expected-4)*COIN)
+
+        # Ensure the change from that transaction appears
+        tx = self.nodes[0].getrawtransaction(txid, 1)
+        change_vout = filter(lambda v: v['valueZat'] != 3*COIN, tx['vout'])
+        change = change_vout[0]['scriptPubKey']['addresses'][0]
+        bal = self.nodes[2].getaddressbalance(change)
+        assert(bal['received'] > 0)
+        assert(bal['received'] < (4-3)*COIN)
+        assert_equal(bal['received'], bal['balance'])
+        assert_equal(self.nodes[2].getaddresstxids(change), [txid])
+
+        # Further checks that limiting by height works
+
+        # non-overlapping range returns an empty list
+        height_txids = self.nodes[1].getaddresstxids({
+            'addresses': [a],
+            'start': 106,
+            'end': 105
+        })
+        assert_equal(height_txids, [])
+
+        # various ranges
+        for i in range(5):
+            height_txids = self.nodes[1].getaddresstxids({
+                'addresses': [a],
+                'start': 106,
+                'end': 106+i
+            })
+            assert_equal(height_txids, txids_a[0:i+1])
+
+        height_txids = self.nodes[1].getaddresstxids({
+            'addresses': [a],
+            'start': 0,
+            'end': 108
+        })
+        assert_equal(height_txids, txids_a[0:3])
+        # end=0 means return the entire range
+        height_txids = self.nodes[1].getaddresstxids({
+            'addresses': [a],
+            'start': 107,
+            'end': 0
+        })
+        assert_equal(height_txids, txids_a)
+
+        # Further check specifying multiple addresses
+        txids_all = list(txids_a)
+        txids_all += self.nodes[1].getaddresstxids(addr_p2pkh)
+        txids_all += self.nodes[1].getaddresstxids(addr_p2sh)
+        multitxids = self.nodes[1].getaddresstxids({
+            'addresses': [a, addr_p2sh, addr_p2pkh]
+        })
+        # No dups in return list from getaddresstxids
+        assert_equal(len(multitxids), len(set(multitxids)))
+
+        # set(txids_all) removes its (expected) duplicates
+        assert_equal(set(multitxids), set(txids_all))
+
+        deltas = self.nodes[1].getaddressdeltas({
+            'addresses': [a],
+        })
+        assert_equal(len(deltas), len(expected_deltas))
+        for z in zip(deltas, expected_deltas):
```

Comment:
Love the zip!

However, I think this would be more readable as:
```
        for i in range(len(deltas)):
            assert_equal(deltas[i]['address'],   a)
            assert_equal(deltas[i]['height'],    expected_deltas[i]['height'])
            ....
```